[Triage Process] Allow SIG maintainers to express their requirements

svrnm opened the pull request:
Since #3990 is merged now, this fixes #4083

## Changes

This introduces what has been discussed in #4083: it adds words to the issue management document to define the `maintainer-blocked` and `maintainer-request` labels and the process on how they can be applied by SIG maintainers.

* [x] Related issues #4083, #3990
* [ ] [`CHANGELOG.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md) file updated for non-trivial changes

pellared commented on issue-management.md:
```diff
@@ -59,3 +59,21 @@ Many SIGs track work in the specification repo that is outside of the triage pro
+
+A triager will add the requested label to the issue if those conditions are met. The difference between the 2 labels is as follows:
+
+- `maintainer-blocked` means that there is an issue with the specification that blocks the SIG from implementing an existing part of the specification.
```
What if there is something missing in the specification that blocks the implementation?

reyang replied:
I guess we can use [bug](https://github.com/open-telemetry/opentelemetry-specification/issues?q=is%3Aissue+is%3Aopen+label%3Abug) for things that require fixes. And if maintainers are blocked by things that don't exist yet, those can be tagged as [feature](https://github.com/open-telemetry/opentelemetry-specification/issues?q=is%3Aissue+is%3Aopen+label%feature). (side note: most GitHub repositories make heavy use of these labels, the fact that OTel spec repo doesn't use it frequently is somewhat odd to me)

pellared commented on issue-management.md:
```diff
@@ -59,3 +59,21 @@ Many SIGs track work in the specification repo that is outside of the triage pro
+A triager will add the requested label to the issue if those conditions are met. The difference between the 2 labels is as follows:
+
+- `maintainer-blocked` means that there is an issue with the specification that blocks the SIG from implementing an existing part of the specification.
+- `maintainer-request` means that there is no blocking issue with an existing part of the specification. This means, that the SIG either wants to implement something that is not yet part of the specification, or that they can go forward with the implementation regardless, but want to clarify ambiguity or inaccuracy.
```
> they can go forward with the implementation regardless

This seems very unwanted

reyang replied:
👀

carlosalberto replied:
+1 to what @pellared said. Also, for cases of adding new functionality, there's the OTEP path. Implementation specific doubts (to leave out ambiguity) are fine though, IMHO. But honestly we could use the `maintainer-blocked` label for that purpose too.

pellared commented on issue-management.md:
```diff
@@ -59,3 +59,21 @@ Many SIGs track work in the specification repo that is outside of the triage pro
+A triager will add the requested label to the issue if those conditions are met. The difference between the 2 labels is as follows:
+
+- `maintainer-blocked` means that there is an issue with the specification that blocks the SIG from implementing an existing part of the specification.
+- `maintainer-request` means that there is no blocking issue with an existing part of the specification. This means, that the SIG either wants to implement something that is not yet part of the specification, or that they can go forward with the implementation regardless, but want to clarify ambiguity or inaccuracy.
```
Shouldn't ambiguity be a blocker?

pellared reviewed (commented):
I have a feeling that having a single `maintainer-request` label could be good enough. The TC can prioritize the issue based on their judgement. Different maintelainers may also have different opinions on what is a blocker and what is not.

reyang commented on issue-management.md:
```diff
@@ -59,3 +59,21 @@ Many SIGs track work in the specification repo that is outside of the triage pro
+* they summarize in the issue description what is blocking this issue and/or what change is required to move that issue forward.
+* they tag all maintainers of the implementation repo in the issue description (@open-telemetry/<sig>-maintainers). No action from the other maintainers is expected, except they disagree with
+  the request of this issue being tagged as `maintainer-blocked` or `maintainer-request`
+* they will share the issue with all other SIGs either via the Maintainers Call or via a message to [#otel-maintainers](https://cloud-native.slack.com/archives/C01NJ7V1KRC) on CNCF slack. This way maintainers of other implementation SIGs can comment if they have the same request.
```
```suggestion
* they will share the issue with all other SIGs either via the [Maintainer Meeting](https://github.com/open-telemetry/community?tab=readme-ov-file#cross-cutting-sigs) or via a message to [#otel-maintainers](https://cloud-native.slack.com/archives/C01NJ7V1KRC) on CNCF slack. This way maintainers of other implementation SIGs can comment if they have the same request.
```

reyang replied:
We should also link to https://github.com/open-telemetry/community?tab=readme-ov-file#slack since many folks had trouble joining CNCF Slack.

svrnm commented:
@pellared @reyang thanks for the initial feedback. I tried to distinguish between "blocked" and "requested" to create some kind of "escalation levels" but based on your comments and re-thinking this may not be necessary. I can rewrite it with only one label (`maintainer-request` ?)

reyang commented:
> @pellared @reyang thanks for the initial feedback. I tried to distinguish between "blocked" and "requested" to create some kind of "escalation levels" but based on your comments and re-thinking this may not be necessary. I can rewrite it with only one label (`maintainer-request` ?)

I don't have strong preference between the two proposals.

When I think about the process, I can see the following cases which we might need to consider:
1. If an issue is created by someone who is not a maintainer, later a maintainer saw it and said "Oh, that's exactly the issue that I'm facing right now", we need to have a way to tag it (we can also create a new issue and resolve the original one as duplicated, I don't like this approach though).
2. Issues targeting the "Development" or "Deprecated" part of the spec might have much lower priority than issues against the "Stable" part of the spec.

I understand that defining process is very hard. Nobody likes complicated process, but a big/complex project cannot succeed with clear/good process. THANK YOU for helping us to improve the process/efficiency @svrnm!

pellared commented:
I prefer small incremental changes in processes when these are not totally wrong (and need be rebuild from ground 😉) Thus proposing just one label. It will be easier also for maintainers to use as they won't need to think which label to select.

svrnm commented:
Will follow up as soon as possible!

svrnm commented:
https://github.com/open-telemetry/opentelemetry-specification/pull/4229/commits/20cef8e8c5996f78b93ff05efc7868e5a33d0ad6 addresses the feedback provided:

* Use only one label and with that the wording how they are different
* Make it clear that a maintainer can also use comments on an existing issue to express their needs
* add links to the meetings + cncf slack

svrnm commented:
Can someone take a look at the recent changes I made?

trask reviewed (approved):
I'm supportive of trying this out

jack-berg commented on issue-management.md:
```diff
@@ -59,3 +59,18 @@ Many SIGs track work in the specification repo that is outside of the triage pro
+
+SIGs that implement and document the specification in their respective language may run into issues with the specification,
+that block their implementation, or they may want to request a specification change related to an implementation change. To
+highlight issues of that kind SIG maintainers can request a `maintainer-request` label being added
```
Should we add a component of the language that allows the blocked language to be conveyed? Would allow us to know at a glance if something is blocking multiple languages.
```suggestion
highlight issues of that kind SIG maintainers can request a `maintainer-request:<language>` label being added
```

svrnm replied:
I have mixed feelings about that, while I agree on the goal of having that information stored in the labels, searching by those labels is going to be much more complicated. A middle ground would be to have `maintainer-request` and `maintainer-request:<language>` at the same time. Not perfect, but still allows both

jsuereth replied:
I'm not sure we should restrict to just language maintainers, can we use `<otel-project>` instead?

svrnm replied:
> I'm not sure we should restrict to just language maintainers, can we use `<otel-project>` instead?

I think this is related to @carlosalberto question below (https://github.com/open-telemetry/opentelemetry-specification/pull/4229#discussion_r1821760303), maybe we start with the language SIGs and then expand the scope? I think the only issue I see is that if we go towards maintainer-request:<otel-project> we get an explosion of labels since there is a lot of them. In that case I would prefer to only have `maintainer-request`

jack-berg replied:
> I have mixed feelings about that, while I agree on the goal of having that information stored in the labels, searching by those labels is going to be much more complicated

Ugh.. so silly that github doesn't support anysort of regex or glob pattern matching when filtering by labels. As a result, yes, the search would have to be something like: `label:maintainer-request:java,maintainer-request:dotnet,maintainer-request:<language>`, enumerating all the languages in the ecosystem.

svrnm replied:
I would leave the PR as is and we can iterate on it as needed?

carlosalberto commented on issue-management.md:
```diff
@@ -59,3 +59,18 @@ Many SIGs track work in the specification repo that is outside of the triage pro
+highlight issues of that kind SIG maintainers can request a `maintainer-request` label being added
+to an issue, if the following conditions are met:
+
+* the issue description or a comment by the maintainer points to at least one issue in an implementation repo, of which they are a maintainer.
```
does implementation repo includes contrib repos?

svrnm replied:
good question 🤔 ... spend some minutes to think about that, because I see pros&cons in restriciting it to "core" vs opening up to all other repos as well (might include contrib, but also auto instrumentation, collector, operator, docs, ...). 

Even if I see some personal benefit as docs maintainer to open it up, I lean towards saying "no" and only include core repos **for now**: let's start with that restricted scope, which also makes it easier if we implement @jack-berg's suggestion (https://github.com/open-telemetry/opentelemetry-specification/pull/4229#discussion_r1827410600), but keep the option to expand it to more in the future, wdyt?

svrnm replied:
I would leave the PR as is and we can iterate on it as needed?

carlosalberto replied:
@svrnm SGTM

jsuereth reviewed (approved):
I'm on board to try that out.  I think there's some unideal fallout due to how github works, but we can figure out how to work around that.

svrnm commented:
I resolved the still open conversations since they were all outdated.

svrnm commented:
@open-telemetry/technical-committee what's missing to merge this?

reyang commented:
> thanks @reyang
> 
> Who can add the label to the repo?

I've added https://github.com/open-telemetry/opentelemetry-specification/labels/maintainer-request
